src/Ryujinx.Ui.Common/App/ApplicationData.cs

```diff
@@ -19,5 +30,129 @@ public class ApplicationData
+
+                    Nca nca = new(virtualFileSystem.KeySet, ncaFile.Get.AsStorage());
+
+                    if (nca.Header.ContentType != NcaContentType.Program)
+                    {
+                        continue;
+                    }
+
+                    int dataIndex = Nca.GetSectionIndexFromType(NcaSectionType.Data, NcaContentType.Program);
+
+                    if (nca.Header.GetFsHeader(dataIndex).IsPatchSection())
+                    {
+                        patchNca = nca;
+                    }
+                    else
+                    {
+                        mainNca = nca;
+                    }
+                }
+            }
+            else if (extension == ".nca")
+            {
+                mainNca = new Nca(virtualFileSystem.KeySet, file.AsStorage());
+            }
+
+            if (mainNca == null)
+            {
+                Logger.Error?.Print(LogClass.Application,
+                    "Extraction failure. The main NCA was not present in the selected file");
+
+                return string.Empty;
+            }
+
+            (Nca updatePatchNca, _) = ApplicationLibrary.GetGameUpdateData(virtualFileSystem,
+                mainNca.Header.TitleId.ToString("x16"), 0, out _);
+
+            if (updatePatchNca != null)
+            {
+                patchNca = updatePatchNca;
+            }
+
+            IFileSystem codeFs = null;
+
+            if (patchNca == null)
+            {
+                if (mainNca.CanOpenSection(NcaSectionType.Code))
+                {
+                    codeFs = mainNca.OpenFileSystem(NcaSectionType.Code, IntegrityCheckLevel.ErrorOnInvalid);
+                }
+            }
+            else
+            {
+                if (patchNca.CanOpenSection(NcaSectionType.Code))
+                {
+                    codeFs = mainNca.OpenFileSystemWithPatch(patchNca, NcaSectionType.Code,
+                        IntegrityCheckLevel.ErrorOnInvalid);
+                }
+            }
+
+            if (codeFs == null)
+            {
+                Logger.Error?.Print(LogClass.Loader, "No ExeFS found in NCA");
+
+                return string.Empty;
+            }
+
+            const string mainExeFs = "main";
+
+            if (!codeFs.FileExists($"/{mainExeFs}"))
+            {
+                Logger.Error?.Print(LogClass.Loader, "No main binary ExeFS found in ExeFS");
+
+                return string.Empty;
+            }
+
+            using var nsoFile = new UniqueRef<IFile>();
+
+            codeFs.OpenFile(ref nsoFile.Ref, $"/{mainExeFs}".ToU8Span(), OpenMode.Read).ThrowIfFailure();
+
+            NsoReader reader = new NsoReader();
+            reader.Initialize(nsoFile.Release().AsStorage().AsFile(OpenMode.Read)).ThrowIfFailure();
+
+            const int buildIdSize = 16;
+
+            return BitConverter.ToString(reader.Header.ModuleId.ItemsRo.ToArray()).Replace("-", "").ToUpper()
+                [..buildIdSize];
```

Comment:
```suggestion
            return BitConverter.ToString(reader.Header.ModuleId.ItemsRo.ToArray()).Replace("-", "").ToUpper()[..16];
```
You can remove `buildIdSize` it's not really needed.